### DISCUSSION TITLE
How to implement physical model as prior for mean function

### DISCUSSION THREAD

**Discussion by huerg** [cornellius-gp/gpytorch#1777] [Q&A] [answered]
After looking at the issues #674 and #1669 and using what I thought might help I am still uncertain about how to implement a custom mean function in gpytorch (to which I am new). I decided to start with a simple linear mean function derived from a linear regression as an example. Later it should encode, for example, a physical model.

Even though my question concentrates on the class “LinearRegressionMean” I am attaching a fully functional code to help clarify the whole question. Since I am not familiar with the usual way of doing it, please let me know if a shorter one would be better.

( I want to implement a custom mean function since they are, in my humble opinion, easier to model than kernels when one has a physical model describing approximate relationships but wants to make the model better with the help of data, falling back to the analytical function at the data gaps). 

Many thanks!
Luis

[custom_mean.zip](https://github.com/cornellius-gp/gpytorch/files/7268327/custom_mean.zip)

**Comment by lukasfro** [comment, marked as answer]
Hi @huerg, a linear mean function is already provided [here](https://github.com/cornellius-gp/gpytorch/blob/master/gpytorch/means/linear_mean.py). Hope this helps :) Let me know if you have further questions.

**Reply by huerg** [reply]
Dear @lukasfro , thank you for your help. I was wondering whether my question had been formulated in an unpolite manner since no one looked at it for weeks. 
If I understand it right the code snippet below tells me that the regression coefficients are getting added to the graph and updated via gradient descent.  

self.register_parameter(name="weights", parameter=torch.nn.Parameter(torch.randn(*batch_shape, input_size, 1)))
        if bias:
            self.register_parameter(name="bias", parameter=torch.nn.Parameter(torch.randn(*batch_shape, 1)))

Since I want to use an external model I will adapt the code and let you know if I get stuck.
Thank you!
